### ISSUE TITLE
to_json formatting regression

### ISSUE THREAD

**Issue by grigorescu** [zeek/zeek#596]
Been tracking down a JSON formatting problem in `to_json`. The simplest test case I can find is a string from one of the values in `SSL::ct_logs`, `"\x07\xd4\xb7o"`

In Bro 2.6.3, the result of `print to_json(l)` is `"\\x07\\xd4\\xb7o"`. Each backslash preceding the hex-escape is escaped.

http://try.bro.org/#/trybro/saved/355506

In Zeek master, the result of `print to_json(l)` is `"\\x07\xd4\xb7o"`. The backslashes stop being double-escaped.

http://try.bro.org/#/trybro/saved/355504

This is invalid JSON (not 100% sure why, but checked it with jsonlint.com) and causing some parsers to choke.

**Comment by jsiwek** [comment]
In Zeek 3.0, we wanted JSON formatting to not escape valid UTF-8 encodings by default.  And that's basically what happens here: `to_json()` is returning a valid JSON string, but `print` is adding back in escapes.

Can see an example of what I mean with this script:

```zeek
local j = to_json("\x07\xd4\xb7o");

local f = open("myfile");
enable_raw_output(f);
local o = open("/dev/stdout");
enable_raw_output(o);

print j;
print o, j;
print f, j;
```

The raw string data returned by `to_json` is actually `"\\x07Էo"`, and that's ok since JSON spec says:

> A string begins and ends with
> quotation marks.  All Unicode characters may be placed within the
> quotation marks except for the characters that must be escaped:
> quotation mark, reverse solidus, and the control characters (U+0000
> through U+001F).

So I'd suggest we change `print` to pass through valid UTF-8 encodings instead of escape them.

@grigorescu do you have a more specific example/context in which you're using `to_json` other than just `print` in case there's actually more going on or other places to change ?

**Comment by sethhall** [comment]
Modifying print to support utf-8 is totally the right answer!  Most terminals these days can handle UTF-8 and as we move toward supporting it, it's a little more surprising to have print not handle it correctly than anything.

**Comment by grigorescu** [comment]
> @grigorescu do you have a more specific example/context in which you're using to_json other than just print in case there's actually more going on or other places to change ?

I'm using `to_json` in combination with ActiveHTTP; the print was just for debugging. I believe that ActiveHTTP works as intended.

I think UTF-8 makes sense.

**Comment by timwoj** [comment]
So just so I'm clear, this is an issue with `print` not outputting utf-8 correctly? `to_json` is fine and does the right thing?

**Comment by jsiwek** [comment]
@timwoj yes, looked like `to_json()` returned proper sequence of bytes, but `print` adds escapes where really it doesn't have to

**Comment by timwoj** [comment]
I changed the title of this GHI to match the actual issue.

**Comment by timwoj** [comment]
Should control characters continue to be escaped? I'm thinking in the case of a few of our tests such as https://github.com/zeek/zeek/blob/master/testing/btest/Baseline/bifs.fmt/out (has newlines and tabs being escaped), https://github.com/zeek/zeek/blob/master/testing/btest/Baseline/bifs.hexdump/out (has a newline being escaped), and a number of others with similar issues.

**Comment by timwoj** [comment]
The "number of others" above are mostly places where we print out something like a signature. Printing UTF-8 characters in those cases seems like a step backwards. We're already printing out simple ASCII characters in those cases, but printing UTF-8 actually makes it less readable.

**Comment by jsiwek** [comment]
> Should control characters continue to be escaped?

I initially want to say "no" to be more similar to other languages: just output the bytes and I'll explicitly escape things when I want.  But generally hard to say what all the expectations/usages of `print` there are out there and what we'll break by changing it.

In any case, I'm starting to think any change to the default `print` behavior (even just passing through UTF-8) may best target 3.1 rather than 3.0.1 on grounds it's potentially a breaking change (e.g. external plugin btests failing between 3.0.0 and 3.0.1 because of a change in `print` behavior violates stability expectations).

Alternative idea to changing default `print` behavior could be adding in new print modes, configurable either by global option or explicit syntax like `print/<mode>`.

Interested to hear other opinions/approaches.

**Comment by timwoj** [comment]
I agree with this not being a 3.0.1 issue. It's not breaking anything for print to be over-escaping the UTF-8 characters, and there's clearly some other factors in question.

**Comment by timwoj** [comment]
I think the primary problem is that the change modifies `BroString::Render` which is called from lots of other places than just `print` (via `ODesc::Add`). How about adding another flag to this block https://github.com/zeek/zeek/blob/master/src/BroString.h#L75-L84 that controls whether UTF-8 gets escaped, using that as the default, and then overriding it for `print`? `ODesc` already has a flag to enable utf8 support, which is set to false by default, but it doesn't affect the call to `BroString::Render`.

**Comment by jsiwek** [comment]
> How about adding another flag to this block https://github.com/zeek/zeek/blob/master/src/BroString.h#L75-L84 that controls whether UTF-8 gets escaped, using that as the default, and then overriding it for `print`?

Yeah, either that or maybe just having `ODesc` do its own formatting logic with underlying data instead of even calling `BroString::Render`.  No strong opinion/intuition though.

**Comment by sethhall** [comment]
I think the argument to target 3.1 with this change makes sense.

**Comment by timwoj** [comment]
Finally coming back around to this one. After spending a little bit of time looking at it, I think adding a flag to `BroString::render_style` and then making sure that's set in `ODesc` when UTF8 is enabled is a good way to fix this. The thing I'm running into right now is how this gets enabled. 

Do we want it to just always be enabled for `PrintStmt` when the output isn't marked as raw? Like I said above, that would affect a number of tests that output raw byte data for fields like signatures (see `scripts.base.protocols.dns.dnskey` for a good example). What would be truly ideal is if individual `Val`s could mark themselves as raw, and `ODesc` would know not to output those in UTF8 mode. Is anything like that current possible?

**Comment by jsiwek** [comment]
> What would be truly ideal is if individual `Val`s could mark themselves as raw

Alternatively, looks like what's really happening in the particular case is that `to_json()` explicitly knows  the `StringVal` it is returning is a byte sequence that is UTF8 encoded.  Any other code that then goes to output that `StringVal`, like `print` statements, can decide it's fine to emit that byte sequence directly instead of escaping the non-printable-ascii-bytes within it.

The general problem is, traditionally, all other Zeek `string` values are technically binary data without knowledge of any potential encoding, but it's safe and convenient to `print` the printable-ascii-bytes within them.  Python does similar for the `bytes` type.  In fact, if we were to translate the `to_json()` situation to Python, we would say it's returning a `bytes` object and if Python were to `print` it, we'd see similar behavior to what Zeek currently does -- it escapes all bytes except printable-ascii ones.  In that sense, one may argue Zeek's `print` behavior is fine the way it is and, if we change the default to emit arbitrary data that appears to be UTF8 without actually knowing, we risk breaking things that rely on the escaped output for no useful reason, e.g. other people's test baselines, or at least making things less readable than before.

So options I'm thinking:
1) Allow storing the known-encoding-type within `StringVal` whenever we do know it (likely few places at the moment, maybe in a bunch of others we could "guess" it's UTF8) and allow `print` or other logic that goes to output it decide what to do based on that.  But sure how you'd want to handle (non-)propagation of that flag for various string operations, like substrings or concatenations. Seems like trouble.
2) Do the more comprehensive separation of a Text and Binary data types like Python has with `str` and `bytes`.  If we magically had this originally, I guess `to_json()` would have returned a Text object instead and then let whatever output logic decide how to encode it (e.g. maybe based on `LANG`).
3) Do nothing (for now, but maybe plan to undertake option (2) eventually)

Or maybe I made it more complicated/conservative than it needs to be.  Let me know :)

**Comment by timwoj** [comment]
> Or maybe I made it more complicated/conservative than it needs to be

I don't think you did, but I agree I don't think there's an easy path. There's a second problem in that we don't want all of the output `StringVal` from `to_json` to be marked as UTF-8, only the individual `StringVal` elements that should be human-readable. This problem should be fixed by (2) since we could set the individual record fields to be bytes or strings, and handle them accordingly when building the JSON.

This seems like a much larger task now than at first glance. Do we want to move this out of 3.1 and into some future release?

**Comment by jsiwek** [comment]
> Do we want to move this out of 3.1 and into some future release?

Yeah, I'll just remove from a release milestone w/ suggestion we revisit later as part of some larger game-plan/review of Text Handling in general.

**Comment by timwoj** [comment]
There's been a lot of work around printing/escaping UTF8 recently. I wonder if this could be more easily-resolved now.

**Comment by MP-Corelight** [comment]
Need to re-validate...

**Comment by timwoj** [comment]
This is still an issue. The script in https://github.com/zeek/zeek/issues/596#issuecomment-535171536 continues to do the same thing as it always has, escaping the output when it shouldn't. I do think the suggestions in https://github.com/zeek/zeek/issues/596#issuecomment-565684009 still hold though. It would have been ideal to split `string` and `bytes` a long time ago as suggested there in (2), but perhaps (1) is an easy enough suggestion at this point.
